### ISSUE TITLE
Live thumbnails

### ISSUE THREAD

**Issue by Kommynct** [H3rmt/hyprswitch#26]
Since hyprland has single-window capture support, there should be a way to get live thumbnails, which would be quite nice

**Comment by H3rmt** [comment]
I don't know any way to obtain the currently rendered output of a window under Wayland

**Comment by Kommynct** [comment]
I was under the impression that single window capture with a restore token would be enough to enable this, sorry if that was incorrect

**Comment by H3rmt** [comment]
I don't know anything about capturing windows under Wayland. I always thought that you need explicit confirmation by the user for each window to capture it.
Of course, it would be cool to display a live preview, but this should happen without the need for any user confirmation.
If you have any ideas on how to achieve this, I would greatly appreciate your help.

**Comment by SergioRibera** [comment]
Taking screenshots would be very complicated due to the issue of the HyprSwitch application overlaying, and in the screenshot, it would align with the overlaid app, for taking screenshots in Wayland, I use this [libwayshot](https://github.com/waycrate/wayshot)

**Comment by littleblack111** [comment]
i think a hyprland plugin will do the trick. Add a hook on render?

**Comment by H3rmt** [comment]
This project can't be used as a Hyprland plugin, as they need to be written in C++ (as far as I know).

If writing plugins in rust will be possible I would happily transform this into a plugin. (this would also remove some problems with the current way of configuring keybinds)

**Comment by littleblack111** [comment]
no, i meant like as a extention, no need to convert the whole project, just add a plugin for this function to work, if cant connect to that plugin then fallback to normal icon thumnail

**Comment by simplemachine-zz** [comment]
Any progress on this?

**Comment by H3rmt** [comment]
Nope, I still don't know how I would implement that, and as I myself wouldn't use this feature I did no further research.

If anyone knows how to do that or implements it I would of course merge it and add a config value to toggle between icon and real window

**Comment by Kommynct** [comment]
hyprland_toplevel_export should help!

kde does it via pipewire, not sure if that works for this though

**Comment by H3rmt** [comment]
> no, i meant like as a extention, no need to convert the whole project, just add a plugin for this function to work, if cant connect to that plugin then fallback to normal icon thumnail

I am currently building exactly that.

It is currently only used to capture key events but could also be used for that purpose

**Comment by jmanc3** [comment]
@H3rmt  These two programs use that top level protocol to capture screenshots of windows

this one is in c and pretty short
https://github.com/yz9551/hypr-window-picker

this one is rust
https://github.com/WhySoBad/hyprland-preview-share-picker

**Comment by zosy** [comment]
we can take a different approach: when pressing Alt+Tab, make all windows float and arrange them in a row. Their sizes can be adjusted according to the screenwidth/app num. When the shortcut is released, restore all windows to their original positions and focus on the selected one. This way, there’s no performance issue, and the implementation would require very little code.

**Comment by H3rmt** [comment]
hyprshell now uses a plugin. 
I think it should be possible to extract the top surface from all windows from the plugin and send them to hyprshell

**Comment by 2zs0** [comment]
Niri has implemented the live thumbnail switcher feature with native support. Maybe we can take inspiration from it https://github.com/YaLTeR/niri/commit/933ffcb229e9e678b271d4043b1d4d5e2b6fa073
